xdsl/ir.py

```diff
@@ -452,17 +451,22 @@ def create(cls: type[OpT],
                regions: Sequence[Region] | None = None) -> OpT:
         op = Operation.with_result_types(cls, operands, result_types,
                                          attributes, successors, regions)
-        return cast(OpT, op)
+        return op
 
     @classmethod
-    def build(cls: type[OpT],
-              operands: list[Any] | None = None,
-              result_types: list[Any] | None = None,
-              attributes: dict[str, Any] | None = None,
-              successors: list[Any] | None = None,
-              regions: list[Any] | None = None) -> OpT:
+    def build(
+        cls: type[OpT],
+        operands: list[SSAValue | Operation | list[SSAValue | Operation]]
+        | None = None,
+        result_types: list[Any] | None = None,
+        attributes: dict[str, Attribute | Any] | None = None,
+        successors: list[Block] | None = None,
+        regions: list[Region | list[Block] | list[Operation]] | None = None
+    ) -> OpT:
         """Create a new operation using builders."""
-        ...
+        raise NotImplementedError("{}.build(...) must either be implemented by"
```

Comment:
IMO, this is a case where I would prefer an `assert False` or a `pass`. Users of the framework "should" never encounter this behavior, so why are we giving proper exceptions like we would if this was public API?

I guess this comes down to one of my fundamental issues with python, namely that there is no way to stop people from building abstract classes and that there actually is no notion of interfaces. From what I know, it is common to just implement abstract methods with a `pass` as their method body, but I am happy to learn how to do this properly by someone more knowledgeable on python than I am.

Reply:
Actually, I think we want to use the `@abstractmethod`-decorator here. Not sure how well that works with our `irdl_op_definiton` decorator though...